File: python/navigation/global_planning/search/theta_star.py
"""Theta* — any-angle path planning on grids.

Nash, Daniel, Koenig & Felner (2007). Like A*, but on each relaxation it tries a
straight line-of-sight shortcut from the expanded node's parent (path 2), so the
returned path follows true straight segments instead of being locked to grid
edges. The heuristic is Euclidean (straight-line), consistent with the any-angle
g-values; w > 1 is weighted Theta* (Pohl 1970).

Does NOT subclass ``_BestFirstSearch``: that skeleton hardcodes the octile
heuristic and standard grid relaxation, both wrong for any-angle. Reconstruction
and finish-emit are shared via ``_discrete``.
"""

from __future__ import annotations

import heapq
import itertools
import math
import time

from navigation.core.capabilities import Capability, LineOfSightSpace
from navigation.core.params import ParamSet
from navigation.core.planner import GlobalPlanner
from navigation.core.trace import TraceRecorder
from navigation.core.types import Cell, PlanResult, PlanStats

from ._discrete import emit_finish, reconstruct


def _euclid(a: Cell, b: Cell) -> float:
    # Euclidean distance on integer cell-index deltas. math.sqrt — NOT math.hypot:
    # hypot is relaxed-accuracy and can differ 1 ULP from the C++ run, diverging
    # f-values / the trace stream the bench compares across languages. sqrt is
    # correctly-rounded, and sqrt(2.0) exactly equals the diagonal edge cost from
    # neighbors(), so a diagonal shortcut (path 2) and a diagonal step (path 1)
    # are bit-equal.
    dr = float(a[0] - b[0])
    dc = float(a[1] - b[1])
    return math.sqrt(dr * dr + dc * dc)


class ThetaStar(GlobalPlanner[Cell, "LineOfSightSpace[Cell]"]):
    def __init__(self, params: ParamSet) -> None:
        super().__init__(params)
        self._heuristic_weight = params.get_float("heuristic_weight")

    @property
    def name(self) -> str:
        return "theta_star"

    def required_capabilities(self) -> set[Capability]:
        return {Capability.DISCRETE_SPACE, Capability.LINE_OF_SIGHT_SPACE}

    def plan(
        self,
        space: LineOfSightSpace[Cell],
        start: Cell,
        goal: Cell,
        recorder: TraceRecorder | None = None,
    ) -> PlanResult[Cell]:
        t0 = time.monotonic()
        w = self._heuristic_weight

        def h(c: Cell) -> float:
            return w * _euclid(c, goal)

        g: dict[Cell, float] = {start: 0.0}
        parent: dict[Cell, Cell] = {start: start}  # self-parent: grandparent + reconstruct stop
        closed: set[Cell] = set()
        counter = itertools.count()  # stable tie-break, keeps heap entries comparable
        frontier: list[tuple[float, int, Cell]] = [(h(start), next(counter), start)]
        expanded = 0
        found = False
        while frontier:
            _, _, s = heapq.heappop(frontier)
            if s in closed:
                continue
            closed.add(s)
            expanded += 1
            if recorder is not None:
                recorder.node_expanded(s, g[s])
            if s == goal:
                found = True
                break
            p = parent[s]
            for s2, edge_cost in space.neighbors(s):
                if s2 in closed:
                    continue
                if space.line_of_sight(p, s2):
                    # Path 2 — any-angle shortcut straight from the grandparent.
                    ecost = _euclid(p, s2)
                    cand = g[p] + ecost
                    par = p
                else:
                    # Path 1 — standard grid step through s.
                    ecost = edge_cost
                    cand = g[s] + ecost
                    par = s
                if s2 not in g or cand < g[s2]:
                    g[s2] = cand
                    parent[s2] = par
                    if recorder is not None:
                        recorder.candidate_evaluated(s2, cand)
                        recorder.edge_added(s2, par, ecost)
                    heapq.heappush(frontier, (cand + h(s2), next(counter), s2))

        runtime = time.monotonic() - t0
        if not found:
            emit_finish(recorder, False, [], 0.0, expanded, runtime)
            return PlanResult(success=False, stats=PlanStats(expanded_nodes=expanded))
        path = reconstruct(parent, start, goal)
        cost = g[goal]  # any-angle path cost; adjacent-edge summing would be wrong for jumps.
        emit_finish(recorder, True, path, cost, expanded, runtime)
        return PlanResult(True, path, cost, PlanStats(expanded_nodes=expanded))

File: cpp/src/global_planning/search/theta_star.cpp
#include "navigation/global_planning/search/theta_star.hpp"

#include <chrono>
#include <cmath>
#include <queue>
#include <unordered_map>
#include <vector>

#include "navigation/global_planning/search/discrete_search.hpp"

namespace navigation::global_planning {

using core::LineOfSightSpace;

namespace {

// Euclidean distance on integer cell-index deltas (row, col). sqrt — NOT hypot:
// hypot is relaxed-accuracy and can differ 1 ULP between C++/Python, which would
// diverge f-values and the emitted trace stream the bench compares across
// languages. sqrt is correctly-rounded, and sqrt(2.0) exactly equals the
// diagonal edge cost from neighbors(), so a diagonal shortcut (path 2) and a
// diagonal grid step (path 1) are bit-equal.
double euclid(const Cell& a, const Cell& b) {
  double dr = static_cast<double>(a.row - b.row);
  double dc = static_cast<double>(a.col - b.col);
  return std::sqrt(dr * dr + dc * dc);
}

}  // namespace

core::PlanResult<Cell> ThetaStarPlanner::plan(LineOfSightSpace<Cell>& space, const Cell& start,
                                              const Cell& goal, TraceRecorder* recorder) {
  // Weighted Theta*: f = g + w*h, h Euclidean (straight-line). w == 1 is standard
  // Theta*; the octile heuristic (space.heuristic) would be inadmissible for
  // any-angle costs since octile >= Euclidean (Nash, Daniel, Koenig & Felner 2007;
  // Pohl 1970).
  const double w = params_.get_float("heuristic_weight");
  auto h = [&](const Cell& c) { return w * euclid(c, goal); };

  struct QItem {
    double f;
    unsigned long long seq;  // insertion order; FIFO tie-break keeps ties stable
    Cell cell;
  };
  struct Greater {
    bool operator()(const QItem& a, const QItem& b) const {
      // Same (f, seq) tie-break as A* so both languages settle the same path.
      return a.f > b.f || (a.f == b.f && a.seq > b.seq);
    }
  };

  auto t0 = std::chrono::steady_clock::now();
  std::priority_queue<QItem, std::vector<QItem>, Greater> open;
  std::unordered_map<Cell, double> g;
  std::unordered_map<Cell, Cell> parent;
  std::unordered_map<Cell, bool> settled;
  unsigned long long seq = 0;

  g[start] = 0.0;
  parent[start] = start;  // self-parent: grandparent lookup + reconstruct terminate
  open.push({h(start), seq++, start});

  PlanResult<Cell> result;
  PlanStats stats;
  bool found = false;

  while (!open.empty()) {
    Cell s = open.top().cell;
    open.pop();
    if (settled[s]) continue;  // stale duplicate with a worse key
    settled[s] = true;
    if (recorder) recorder->node_expanded(s, g[s]);
    ++stats.expanded_nodes;
    if (s == goal) {
      found = true;
      break;
    }
    const Cell p = parent[s];
    for (const auto& [s2, edge_cost] : space.neighbors(s)) {
      if (settled[s2]) continue;
      double cand;
      Cell par;
      double ecost;
      if (space.line_of_sight(p, s2)) {
        // Path 2 — any-angle shortcut straight from the grandparent.
        ecost = euclid(p, s2);
        cand = g[p] + ecost;
        par = p;
      } else {
        // Path 1 — standard grid step through s.
        ecost = edge_cost;
        cand = g[s] + ecost;
        par = s;
      }
      auto it = g.find(s2);
      if (it == g.end() || cand < it->second) {
        g[s2] = cand;
        parent[s2] = par;
        if (recorder) recorder->candidate_evaluated(s2, cand);
        if (recorder) recorder->edge_added(s2, par, ecost);
        open.push({cand + h(s2), seq++, s2});
      }
    }
  }

  if (found) {
    result.success = true;
    result.path = reconstruct_path(parent, start, goal);
    result.cost = g[goal];  // NOT path_cost(): it sums adjacent edges only and is ~0 for jumps.
    if (recorder) recorder->path_found(result.path);  // emit_finished_discrete does not.
  }
  result.stats = stats;
  double rt = std::chrono::duration<double>(std::chrono::steady_clock::now() - t0).count();
  emit_finished_discrete(recorder, result.success, result.cost, stats, rt);
  return result;
}

}  // namespace navigation::global_planning
